File: leetcode/91-decode-ways.py
# https://leetcode.com/problems/decode-ways/ (top-interview-questions)

class Solution:
    def numDecodings(self, s: str) -> int:
        cache = dict()
        
        def dfs(i):
            if i == len(s):
                return 1
            if i in cache:
                return cache[i]
            if s[i] == '0':
                return 0
            
            result = dfs(i+1) # in this recursion call we try to split the number into part of single digit
            # here we see if we can to split the number into part of two digits
            if i+1 < len(s) and s[i:i+2] < '27':
                result += dfs(i+2)

            cache[i] = result
            return result
        
        return dfs(0)

File: leetcode/91-decode-ways.js
// https://leetcode.com/problems/decode-ways/ (top-interview-questions)

/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function(s) {
  const cache = {};

  const dfs = (i) => {
    if (i === s.length) {
      return 1;
    }

    // string that start with 0 is invalid - so we return 0 ( 0 decode way )
    if (s.charAt(i) === '0') {
      return 0;
    }

    if (cache.hasOwnProperty(i)) {
      return cache[i];
    }

    let result = dfs(i+1);

    if (i+1 < s.length && Number(s.slice(i, i+2)) < 27) { // if ( s[i..i+1] between 9 and 27 )
      result += dfs(i+2);
    }

    cache[i] = result;
    return result;
  }

  return dfs(0);
};
